pkg/storegateway/storegatewaypb/custom.go

```diff
@@ -0,0 +1,146 @@
+	if err != nil {
+		return client, wrapContextError(err)
+	}
+
+	return newCustomSeriesClient(client), nil
+}
+
+// LabelNames implements StoreGatewayClient.
+func (c *customStoreGatewayClient) LabelNames(ctx context.Context, in *storepb.LabelNamesRequest, opts ...grpc.CallOption) (*storepb.LabelNamesResponse, error) {
+	res, err := c.wrapped.LabelNames(ctx, in, opts...)
+	return res, wrapContextError(err)
+}
+
+// LabelValues implements StoreGatewayClient.
+func (c *customStoreGatewayClient) LabelValues(ctx context.Context, in *storepb.LabelValuesRequest, opts ...grpc.CallOption) (*storepb.LabelValuesResponse, error) {
+	res, err := c.wrapped.LabelValues(ctx, in, opts...)
+	return res, wrapContextError(err)
+}
+
+// customStoreGatewayClient is a custom StoreGateway_SeriesClient which wraps well known gRPC errors into standard golang errors.
+type customSeriesClient struct {
+	*customClientStream
+
+	wrapped StoreGateway_SeriesClient
+}
+
+func newCustomSeriesClient(client StoreGateway_SeriesClient) *customSeriesClient {
+	return &customSeriesClient{
+		customClientStream: &customClientStream{client},
+		wrapped:            client,
+	}
+}
+
+func (c *customSeriesClient) Recv() (*storepb.SeriesResponse, error) {
+	res, err := c.wrapped.Recv()
+	return res, wrapContextError(err)
+}
+
+// customClientStream is a custom grpc.ClientStream which wraps well known gRPC errors into standard golang errors.
+type customClientStream struct {
+	wrapped grpc.ClientStream
+}
+
+// Header implements grpc.ClientStream.
+func (c *customClientStream) Header() (metadata.MD, error) {
+	md, err := c.wrapped.Header()
+	return md, wrapContextError(err)
+}
+
+// Trailer implements grpc.ClientStream.
+func (c *customClientStream) Trailer() metadata.MD {
+	return c.wrapped.Trailer()
+}
+
+// CloseSend implements grpc.ClientStream.
+func (c *customClientStream) CloseSend() error {
+	//nolint:forbidigo // Here we're just wrapping CloseSend() so it's OK to call it.
+	return wrapContextError(c.wrapped.CloseSend())
+}
+
+// Context implements grpc.ClientStream.
+func (c *customClientStream) Context() context.Context {
+	return c.wrapped.Context()
+}
+
+// SendMsg implements grpc.ClientStream.
+func (c *customClientStream) SendMsg(m any) error {
+	return wrapContextError(c.wrapped.SendMsg(m))
+}
+
+// RecvMsg implements grpc.ClientStream.
+func (c *customClientStream) RecvMsg(m any) error {
+	return wrapContextError(c.wrapped.RecvMsg(m))
+}
+
+func wrapContextError(err error) error {
+	switch {
+	case err == nil:
+		return nil
+	case grpcutil.ErrorToStatusCode(err) == codes.Canceled:
+		return &grpcContextError{grpcErr: err, stdErr: context.Canceled}
+	case grpcutil.ErrorToStatusCode(err) == codes.DeadlineExceeded:
+		return &grpcContextError{grpcErr: err, stdErr: context.DeadlineExceeded}
+	default:
+		return err
+	}
+}
+
+// grpcContextError is a custom error used to wrap gRPC errors which maps to golang standard context errors.
+type grpcContextError struct {
```

Comment:
Should we make this error implement `GRPCStatus() *Status` method? `status.FromError` from grpc-go will work correct with this type, because it uses `errors.As` unwrapping, but `status.FromError` from "gogo" package does not. Our code base uses this call as well (in separate PR, we could disable its use, and only use `status.FromError` from grpc-go library.)

Reply:
What do you think about this approach https://github.com/grafana/mimir/pull/6934/commits/a2688aeb12e974afbff1993e9eef137ae73d7303 ?

Reply:
It looks good. Thank you.